kong/plugins/cors/handler.lua

```diff
@@ -1,88 +1,143 @@
 local BasePlugin = require "kong.plugins.base_plugin"
-local responses = require "kong.tools.responses"
+local responses  = require "kong.tools.responses"
+
+
+local req_get_headers = ngx.req.get_headers
+local req_get_method  = ngx.req.get_method
+local re_find         = ngx.re.find
+local concat          = table.concat
+local tostring        = tostring
+local ipairs          = ipairs
+
 
 local CorsHandler = BasePlugin:extend()
 
+
 CorsHandler.PRIORITY = 2000
 
-local OPTIONS = "OPTIONS"
 
 local function configure_origin(ngx, conf)
-  if conf.origin == nil then
+  if not conf.origins then
     ngx.header["Access-Control-Allow-Origin"] = "*"
-  else
-    ngx.header["Access-Control-Allow-Origin"] = conf.origin
-    ngx.header["Vary"] = "Origin"
+    ngx.ctx.cors_allow_all = true
+    return
+  end
+
+  if #conf.origins == 1 then
+    if conf.origins[1] == "*" then
+      ngx.ctx.cors_allow_all = true
+
+    else
+      ngx.header["Vary"] = "Origin"
+    end
+
+    ngx.header["Access-Control-Allow-Origin"] = conf.origins[1]
+    return
+  end
+
+  local req_origin = req_get_headers()["origin"]
```

Comment:
using `ngx.var` as opposed to `get_headers` would possibly be more performant here